common/retry_delays.go

```diff
@@ -0,0 +1,138 @@
+package common
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/agglayer/aggkit/config/types"
+)
+
+var (
+	// If this error is in the returned error, it means that the function should be aborted
+	// and no more retries should be attempted.
+	ErrAbort          = fmt.Errorf("abort")
+	ErrInvalidConfig  = fmt.Errorf("invalid retry delays config")
+	ErrExecutionFails = fmt.Errorf("execution fails after retries exceeded")
+)
+
+// RetryDelays is a struct that holds the retry delays and the maximum number of retries.
+// It implements the RetryDelayer interface, which allows executing a function with retry logic.
+// The delays are specified as a slice of types.Duration, and the maximum number of retries can be set.
+// If maxRetries is set to 0, it means infinite retries are allowed.
+//
+// You can also abort the retrying wrapping the error ErrAbort into the result
+// this is useful if there are some conditions that should not be retried
+type RetryDelays struct {
+	Delays []types.Duration
+	// MaxAttempts is the maximum number of retries to attempt.
+	// if MaxAttempts is 0, it means infinite retries.
+	// if MaxAttempts is 1, it means no retries will be attempted.
+	MaxAttempts int
+}
+
+func (r *RetryDelays) Validate() error {
+	if r == nil {
+		return fmt.Errorf("retry delays cannot be nil. %w", ErrInvalidConfig)
+	}
+	if len(r.Delays) == 0 {
+		return fmt.Errorf("retry delays cannot be empty. %w", ErrInvalidConfig)
+	}
+	for _, delay := range r.Delays {
+		if delay.Duration <= 0 {
+			return fmt.Errorf("retry delay must be greater than zero, got %s. %w",
+				delay.Duration, ErrInvalidConfig)
+		}
+	}
+	if r.MaxAttempts < 0 {
+		return fmt.Errorf("max retries cannot be negative, got %d. %w",
+			r.MaxAttempts, ErrInvalidConfig)
+	}
+	return nil
+}
+
+func (r *RetryDelays) String() string {
+	if r == nil {
+		return "RetryDelays is nil"
+	}
+	return fmt.Sprintf("RetryDelays{Delays: %v, MaxRetries: %d}", r.Delays, r.MaxAttempts)
+}
+
+// Execute executes the provided function with retry logic.
+// retryDelaysConfig: it's the RetryDelays struct that holds the retry delays and the maximum number of retries.
+// ctx: the context to use for cancellation.
+// logFunc: a function to log messages, if nil a silent logger will be used.
+// name: the name of the operation, used for logging.
+// fn: the function to execute, it should return a result of type T and an error.
+// If the function returns an error that is wrapped with ErrAbort,
+// the execution will be aborted and no more retries will be attempted.
+func Execute[T any](retryDelaysConfig *RetryDelays,
+	ctx context.Context,
+	logFunc func(format string, args ...interface{}),
+	name string,
+	fn func() (T, error)) (T, error) {
+	var zero T
+	if err := retryDelaysConfig.Validate(); err != nil {
+		return zero, err
+	}
+	if logFunc == nil {
+		// if logger is nil, we create a silent logger
+		logFunc = func(format string, args ...interface{}) {}
+	}
+
+	retries := 0
+	for {
```

Comment:
I think you can move the first if to the for loop itself:

```suggestion
	for retryAttempt := 0; retryDelaysConfig.MaxAttempts == 0 || retryAttempt < retryDelaysConfig.MaxAttempts; retryAttempt++
```

Reply:
Yes, good suggestion! 340f317f474bc188d3969c7366f6d0d69851cbfa